scrapy_poet/injection.py

```diff
@@ -142,14 +158,44 @@ def build_instances_from_providers(
             if not provided_classes:
                 continue
 
-            kwargs = andi.plan(
-                provider,
-                is_injectable=is_injectable,
-                externally_provided=scrapy_provided_dependencies,
-                full_final_kwargs=False,
-            ).final_kwargs(scrapy_provided_dependencies)
-            objs = yield maybeDeferred_coro(provider, set(provided_classes),
-                                               **kwargs)
+            objs, fingerprint = None, None
+            cache_hit = False
+            if self.cache:
+                if not provider.name:
+                    raise NotImplementedError(f"The provider {type(provider)} must have a `name` defined if"
+                                              f" you want to use the cache. It must be unique across the providers.")
+                # Return the data if it is already in the cache
+                fingerprint = f"{provider.name}_{provider.fingerprint(set(provided_classes), request)}"
+                try:
+                    data = self.cache[fingerprint]
+                except KeyError:
+                    self.crawler.stats.inc_value("scrapy-poet/cache/miss")
+                else:
+                    self.crawler.stats.inc_value("scrapy-poet/cache/hit")
+                    if isinstance(data, BaseException):
+                        raise data
+                    objs = provider.deserialize(data)
+                    cache_hit = True
+
+            if not objs:
+                kwargs = andi.plan(
+                    provider,
+                    is_injectable=is_injectable,
+                    externally_provided=scrapy_provided_dependencies,
+                    full_final_kwargs=False,
+                ).final_kwargs(scrapy_provided_dependencies)
+                try:
+
+                    # Invoke the provider to get the data
+                    objs = yield maybeDeferred_coro(provider, set(provided_classes), **kwargs)
+
+                except BaseException as e:
```

Comment:
Why is BaseException caught here, not just Exception? It'd include system-exiting exception; I'm not sure we should be catching those.

Reply:
Agreed. Updated on https://github.com/scrapinghub/scrapy-poet/pull/55/commits/1fd5f891d56d92c3070ccc02178b4f86adac74fa